### DISCUSSION TITLE
Can the set of input variables of a BMI-compliant model be defined only at the initialization?

### DISCUSSION THREAD

**Discussion by adlzanchetta** [csdms/bmi#141] [Q&A] [answered]
**Scenario:**

Lets say I have a hydrological model that, depending on its config file, may require one of the following sets of inputs:

1. precipitation;
2. mean hourly temperature.

or

1. precipitation;
2. minimum daily temperature;
3. maximum daily temperature.

or

1. accumulated rainfall;
2. accumulated snowfall;
3. mean daily temperature;
4. mean daily wind speed.

So that I can only determine the return of ```get_input_item_count()``` and ```get_input_var_names()``` AFTER ```initialize()``` is called. I was not able to find in the BMI documentation an explicit assertion that consumers of BMI models are required (or not) to call ```initialize()``` before ```get_input_...``` functions.

**Question:**

Does the BMI standard requires functions like ```get_input_item_count()``` and ```get_input_var_names()``` to be able to return a valid value even BEFORE ```initialize()``` is called? Or can I assume that a whatever uses my model via its BMI interface will ALWAYS call ```initialize()``` before trying to call ```get_input_item_count()``` or ```get_input_var_names()``` ?

**Suggestion:**

One of the following items could be added to the BMI specs (if not already there):

- consumers of BMI models must call ```initialize()``` always before ```get_input_...``` methods (or before ANY other methods); or
- BMI models must be able to return valid values for the methods ```get_input_...``` always, even before ```initialize()```.

I personaly am inclined to the first of them, but there is clear value in both.

**Comment by PhilMiller** [comment]
I can't speak for the standard as such, but as a developer of a consumer/client code, I can say that NOAA-OWP/ngen will call `initialize` before any other routine (in the general case).

If this does end up leading to elaboration in the specification, it's worth noting whether any calls are allowed after `finalize`.

**Comment by mcflugen** [comment, marked as answer]
@adlzanchetta `initialize` should be called before any other BMI functions are called. I'm not sure we've specified that in the docs, but we should. In our framework, we assume that any BMI method called before *initialize* cannot be trusted.

We do have components that have some functions that can be called successfully, and return valid information, even if *initialize* hasn't yet been called, but a framework can't assume that. These functions are typically ones that return static metadata like a model that, for example, always provides a fixed set of output variables.

**Reply by RolfHut** [reply]
I would consider it a good best practice to return a nice error code (depending on programming language error handling) if any another BMI function is called before initialize.

**Comment by adlzanchetta** [comment]
Thank you @PhilMiller and @mcflugen for the clarification.
So we will assume that ```initialize``` will always be called before any other method.
This seems to be an implicit assumption so widely adopted that became part of the _de facto_ standard.
